Provision new jenkins build box

stvstnfrd opened the pull request.

stvstnfrd commented on stanford/roles/cluster/defaults/main.yml:
```diff
@@ -181,6 +195,32 @@ CLUSTER_SECURITY_GROUP_RULES:
+            to_port: 443
+            cidr_ip: '0.0.0.0/0'
+      subnets:
+        - "{{ DEPLOYMENT_SUBNET_PUBLIC }}"
```
FWIW, much of that change (`s/nat/public`) already exists on master [1][2];
we were already calling that subnet `public` and already placing public ELBs within.
The renaming that you see in this PR is me paying down technical debt and synchronizing the AWS tags to match the code itself.

I'm admittedly not clear on what you're being confused by, nor do I follow what you're suggesting.

> Given the amount of time it has taken me to eyeball this PR

Can you clarify what exactly you're having trouble with here?

I had considered this PR to be straight-forward and non-controversial.
Just like every other cluster we've configured to-date, it only includes:
- configuration for:
    - the subnet
    - the elb
    - the security groups
- a new playbook to invoke the cluster

The only 2/3 "unrelated" lines are to create 2 new variables which fix a bug (elbs going to the wrong subnet) and making it easier to extend (explained below).

> if [...] code can be split out to make it more autonomous

As we've discussed f2f, a "cluster" in this EDC architecture provides the infrastructure for a cluster/box/service to operate. These tasks, performed by `roles/cluster/tasks/main.yml` include:
- create a subnet
- add the subnet to a route table
- optionally, create an ELB w/ its own security group
- create a security group for the cluster itself
- if an ec2-backed cluster, create an ec2 instance.

These tasks, unless noted otherwise, are all required.

It's not clear to me what can be simplified; are you seeing something that I'm not?
Excluding `debug` statements (which probably should be removed at this point anyway),
the entire task file is under 100 LoC.

What are you trying to optimize?

I'm open to suggestions, but I'll need some more info on what exactly you're hoping to solve here.

[1] https://github.com/Stanford-Online/configuration/blob/master/stanford/roles/cluster/defaults/main.yml#L24
[2] https://github.com/Stanford-Online/configuration/blob/master/stanford/roles/cluster/defaults/main.yml#L89

dcadams replied:
Sorry, I should have been more explicit.

My understanding is that this pr contains 2 functional changes:
- Adds a cluster for build, aka jenkins.
- Changes the subnet attached to the LB for all clusters that have an LB.

It's the second item that requires considerable perspicacity to trace though because it involves looking at each of the route tables for each of the LB subnets and comparing it to the deployment 'instance' subnet. Isn't that correct?

My, admittedly, oblique reference to autonomous code was that, in this case, it would have been easier if the code for each cluster was self-contained such that this change would only have affected the 'build' subnet. However, that's not an ideal solution, for obvious reasons, so I'm not advocating for that. Just that maybe there's some better way of architecting this.

dcadams replied:
"However, for the public ELBs, the configured subnet was
ignored and we ended up updating it manually."

Well, that was my point; the subnets for these elbs will change. I didn't know that they had been updated manually but even when I look at them now they look odd. E.g. lagunita-elb-edxapp-public has 2 subnets that both have private RTs. It doesn't look to me as though the subnets for this ELB were edited.

dcadams commented on stanford/roles/deployment/defaults/main.yml:
```diff
@@ -6,3 +6,5 @@ DEPLOYMENT_REGION: "{{ region }}"
 DEPLOYMENT_SSL_CERTIFICATE_ID: 'arn:aws:acm:<REGION>:<ACCOUNT>:certificate/<GUID>'
 # DEPLOYMENT_VPC_ID: vpc-deadbeefcafe
```
Let's leave in then.

stvstnfrd replied:
Ok.

dcadams commented on stanford/roles/deployment/tasks/main.yml:
```diff
@@ -34,9 +34,9 @@
     az: "{{ CLUSTER_NAT_SUBNET.zone }}"
     resource_tags:
-      Cluster: nat
+      Cluster: public
```
I actually can't remember that conversation but it being 'nat' is good with me also.

stvstnfrd replied:
The "noun vs. adjective" debate aside, I think calling it `nat` is objectively worse.
Putting a public ELB into a `nat` subnet, vs. putting a public `elb` into a `public` subnet, seems even more opaque and confusing/misleading.

As mentioned, I think the optimal approach would be to fully split out the two (`nat` and `public`), but that requires additional work, as the resources are already co-mingled. I'm open to addressing this work in the future, but i don't think we should be moving unrelated ELBs at this time. The only reason I waded in originally was because the bug prevented the jenkins/build deploy from working properly.

And to reiterate, this is already referred to as `public` and _not_ `nat` throughout each production deployment; I was just updating the code that generates this tag to match the value the tags already had in prod.

You're right that intermingling the terminology can be confusing, so I've gone ahead and pushed up an additional commit that renames any additional references to `nat` (except "Create NAT gateway") to say `public`.

dcadams commented on stanford/roles/deployment/tasks/main.yml:
```diff
@@ -34,9 +34,9 @@
+      Cluster: public
       Deployment: "{{ COMMON_DEPLOYMENT }}"
-      Name: "{{ COMMON_DEPLOYMENT }}-subnet-nat-{{ DEPLOYMENT_NUMBER }}"
+      Name: "{{ DEPLOYMENT_SUBNET_PUBLIC_NAME }}"
```
Not a big deal with me; either way.

stvstnfrd replied:
Okay, let's leave it as-is for now.

caesar2164 commented:
YAML syntax seems good. 👍

stvstnfrd commented:
I rebased and squashed this, then tested once more by merging this branch into my `ginkgo/wip` branch. I thought things were broken at first, but it turned out the box just needed restarted once.
